### DISCUSSION TITLE
Logging TLS messages

### DISCUSSION THREAD

**Discussion by OMyCode** [randombit/botan#3719] [Q&A]
I have been looking into the  the TLS Server/Client/Channel classes to see if they support the logging of structured TLS messages into a file. What I am after is to log all TLS communication for later inspection, debugging and analysis. Any suggestions?

Cheers

**Comment by reneme** [comment]
You can override `Botan::TLS::Callbacks::tls_inspect_handshake_msg()`. This will be called for every message that is received or sent and will be passed a `Botan::TLS::Handshake_Message`, which it the base class of all handshake messages.

Using `Handshake_Message::type()` (or `::type_string()` for a human-readable representation) you can determine the type of the message and then down-cast it into the appropriate concrete message (see [tls_messages.h](https://github.com/randombit/botan/blob/master/src/lib/tls/tls_messages.h)). Or you can call `Handshake_Message::serialize()` to get the wire format of each message.
